Vector assert

nmiell opened the pull request:
gcc's debug std::vector will assert if operator[] indexes outside the vector's size.

I've run into a couple of places where &vector[0] is used unconditionally before iterating over the vector.

baldurk commented on renderdoc/driver/vulkan/vk_shader_cache.cpp:
```diff
@@ -294,7 +294,7 @@ void VulkanShaderCache::MakeGraphicsPipelineInfo(VkGraphicsPipelineCreateInfo &p
   specdata.resize(specSize);
 
-  VkSpecializationMapEntry *entry = &specMapEntries[0];
+  VkSpecializationMapEntry *entry = specEntries ? &specMapEntries[0] : NULL;
```
I'd prefer `specEntries.data()` here which amounts to the same thing (not guaranteed to be `NULL`, but we don't need that here as we don't check for `NULL`.

baldurk reviewed (changes requested):
This change is fine, though I'd prefer to use `std::vector::data()` where appropriate. I've been using that for a while but there may be code written in the `&vector[0]` fashion either from before that or if I just forgot.

nmiell committed:
Fix some std::vector assertion failures

gcc's debug std::vector will assert if you use vector[0] when
vector.size() == 0.
